### PR TITLE
[Sema] Diagnose tautological bounds checks

### LINKED ISSUE #120214: -Wtautological-compare should flag tautological pointer bounds checks
`-Wtautological-compare` should report pointer "bounds checks" patterns that are always true/false because unsigned pointer arithmetic cannot overflow.

```
bool test1(const char *ptr, size_t index) {
  return ptr + index < ptr; // always false
}

bool test2(const char *ptr, size_t index) {
  return ptr + index >= ptr; // always true
}
```

These two are fine though:

```
bool test3(const char *ptr, ssize_t index) {
  return ptr + index < ptr; // unknown (same as index < 0)
}

bool test4(const char *ptr, ssize_t index) {
  return ptr + index >= ptr; // unknown (same as index >= 0)
}
```

Idea from https://github.com/llvm/llvm-project/pull/118472#issuecomment-2548141458.

### PR DESCRIPTION
**Pull request by nikic**
This diagnoses comparisons like `ptr + unsigned_index < ptr` and `ptr + unsigned_index >= ptr`, which are always false/true because addition of a pointer and an unsigned index cannot wrap (or the behavior is undefined).

This warning is intended to help find broken bounds checks (which must be implemented in terms of uintptr_t instead).

Fixes https://github.com/llvm/llvm-project/issues/120214.

### CODE REVIEW

**Review by cor3ntin — COMMENTED** [review]
Generally looks great.

Can you add 
 - a changelog entry
 - a test with C arrays?

File: clang/lib/Sema/SemaExpr.cpp [hunk 1]
```diff
@@ -11786,6 +11786,49 @@ static bool checkForArray(const Expr *E) {
   return D->getType()->isArrayType() && !D->isWeak();
 }
 
+/// Detect patterns ptr + size >= ptr and ptr + size < ptr, where ptr is a
+/// pointer and size is an unsigned integer. Return whether the result is
+/// always true/false.
+static std::optional<bool> isTautologicalBoundsCheck(Expr *LHS, Expr *RHS,
```
Comment (tbaederr) [inline clang/lib/Sema/SemaExpr.cpp:11792]:
```suggestion
static std::optional<bool> isTautologicalBoundsCheck(const Expr *LHS, const Expr *RHS,
```

**Review by cor3ntin — APPROVED** [review]
Thanks!

**Review by hstk30-hw — APPROVED** [review]

**Comment by zmodem** [comment]
Nice, thank you!

**Comment by nathanchance** [comment]
@nikic I noticed your comment on #118472 (the motivator for this change AFAICT):

> `-fwrapv` should already cover pointers

but it does not seem like this warning takes that into account (see the last example below)? I noticed a few instances of this warning in the Linux kernel but it sets `-fno-strict-overflow` for well defined overflow semantics it can depend on because it wants to avoid optimizations like the one that triggered this warning (`-fno-delete-null-pointer-checks` is a similarly used flag), so claiming the result of the checks are always false just does not seem to line up with reality. It seems like if the optimizer is going to respect this check and not remove it, this warning should not trigger either.

```c
int check(const int* foo, unsigned int idx)
{
        return foo + idx < foo;
}
```

```
$ clang-19 -O2 -c test.c

$ llvm-objdump -dr test.o
...
0000000000000000 <check>:
       0: 31 c0                         xorl    %eax, %eax
       2: c3                            retq

$ clang-19 -O2 -fno-strict-overflow -c test.o

$ llvm-objdump -dr test.o
...
0000000000000000 <check>:
       0: 89 f0                         movl    %esi, %eax
       2: 48 8d 0c 87                   leaq    (%rdi,%rax,4), %rcx
       6: 31 c0                         xorl    %eax, %eax
       8: 48 39 f9                      cmpq    %rdi, %rcx
       b: 0f 92 c0                      setb    %al
       e: c3                            retq
```
	
```
$ gcc --version | head -1
gcc (GCC) 14.2.1 20240910

$ gcc -O2 -c test.c

$ llvm-objdump -dr test.o
...
0000000000000000 <check>:
       0: 31 c0                         xorl    %eax, %eax
       2: c3                            retq

$ gcc -O2 -fno-strict-overflow -c test.c

$ llvm-objdump -dr test.o
...
0000000000000000 <check>:
       0: 89 f6                         movl    %esi, %esi
       2: 48 8d 04 b7                   leaq    (%rdi,%rsi,4), %rax
       6: 48 39 f8                      cmpq    %rdi, %rax
       9: 0f 92 c0                      setb    %al
       c: 0f b6 c0                      movzbl  %al, %eax
       f: c3                            retq
```

```
$ clang --version | head -1
ClangBuiltLinux clang version 20.0.0git (https://github.com/llvm/llvm-project.git 99c2e3b78210a345afb1b5121f12b0e7bf923543)

$ clang -O2 -c test.c
test.c:3:19: warning: pointer comparison always evaluates to false [-Wtautological-compare]
    3 |         return foo + idx < foo;
      |                          ^
1 warning generated.

$ llvm-objdump -dr test.o
...
0000000000000000 <check>:
       0: 31 c0                         xorl    %eax, %eax
       2: c3                            retq

$ clang -O2 -c -fno-strict-overflow test.c
test.c:3:19: warning: pointer comparison always evaluates to false [-Wtautological-compare]
    3 |         return foo + idx < foo;
      |                          ^
1 warning generated.

$ llvm-objdump -dr test.o
...
0000000000000000 <check>:
       0: 89 f0                         movl    %esi, %eax
       2: 48 8d 0c 87                   leaq    (%rdi,%rax,4), %rcx
       6: 31 c0                         xorl    %eax, %eax
       8: 48 39 f9                      cmpq    %rdi, %rcx
       b: 0f 92 c0                      setb    %al
       e: c3                            retq
```

I tested something like:

```diff
diff --git a/clang/lib/Sema/SemaExpr.cpp b/clang/lib/Sema/SemaExpr.cpp
index e06a092177ef..2dd7c5951d71 100644
--- a/clang/lib/Sema/SemaExpr.cpp
+++ b/clang/lib/Sema/SemaExpr.cpp
@@ -11789,10 +11789,11 @@ static bool checkForArray(const Expr *E) {
 /// Detect patterns ptr + size >= ptr and ptr + size < ptr, where ptr is a
 /// pointer and size is an unsigned integer. Return whether the result is
 /// always true/false.
-static std::optional<bool> isTautologicalBoundsCheck(const Expr *LHS,
+static std::optional<bool> isTautologicalBoundsCheck(Sema &S,
+                                                     const Expr *LHS,
                                                      const Expr *RHS,
                                                      BinaryOperatorKind Opc) {
-  if (!LHS->getType()->isPointerType())
+  if (!LHS->getType()->isPointerType() || S.getLangOpts().isSignedOverflowDefined())
     return std::nullopt;

   // Canonicalize to >= or < predicate.
@@ -11940,7 +11941,7 @@ static void diagnoseTautologicalComparison(Sema &S, SourceLocation Loc,
                                 << 1 /*array comparison*/
                                 << Result);
     } else if (std::optional<bool> Res =
-                   isTautologicalBoundsCheck(LHS, RHS, Opc)) {
+                   isTautologicalBoundsCheck(S, LHS, RHS, Opc)) {
       S.DiagRuntimeBehavior(Loc, nullptr,
                             S.PDiag(diag::warn_comparison_always)
                                 << 2 /*pointer comparison*/
```

which results in a behavior I would expect:

```
$ clang -fsyntax-only test.c
test.c:3:19: warning: pointer comparison always evaluates to false [-Wtautological-compare]
    3 |         return foo + idx < foo;
      |                          ^
1 warning generated.

$ clang -fsyntax-only -fwrapv test.c
```

If I am missing something, please let me know.

**Comment by nikic** [comment]
@nathanchance You are correct, this warning should indeed respect `-fwrapv`/`-fno-strict-overflow`. Your patch looks reasonable to me as well.

**Comment by nathanchance** [comment]
> @nathanchance You are correct, this warning should indeed respect `-fwrapv`/`-fno-strict-overflow`. Your patch looks reasonable to me as well.

Thanks for the confirmation, I have submitted https://github.com/llvm/llvm-project/pull/120480 for this.

**Pull request merged**